### ISSUE TITLE
Stateful/non `'static` handlers?

### ISSUE THREAD

**Issue by seanpianka** [routerify/routerify#16]
Consider a fairly trivial use-case for an API:

```rust
struct Handler {
    db: database::Client
}

impl Handler {
    async fn user_book_handler(&self, req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let user_name = req.param("userName").unwrap();
        let book_name = req.param("bookName").unwrap();

        let user_id = self.db.get(user_name).unwrap();

        Ok(Response::new(Body::from(format!("User ID: {}", user_id))))
    }

    pub fn router(&self) -> Router<Body, io::Error> {
        Router::builder()
            .get("/users/:id", |request| self.load_user(request))
            .build()
            .unwrap()
    }
}

fn main() {
    let handler = Handler { db: database::Client::new("127.0.0.1:6379") };
    let service = RouterService::new(handler.router()).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}
```

This is not possible, given this compiler error:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
  --> src/main.rs:24:36
   |
24 |                 .get("/users/:id", |request| self.load_user(request))
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 22:9...
  --> src/main.rs:22:9
   |
22 | /         pub fn router(&self) -> Router<Body, io::Error> {
23 | |             Router::builder()
24 | |                 .get("/users/:id", |request| self.load_user(request))
25 | |                 .build()
26 | |                 .unwrap()
27 | |         }
   | |_________^
note: ...so that the types are compatible
  --> src/main.rs:24:36
   |
24 |                 .get("/users/:id", |request| self.load_user(request))
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: expected  `&&users::Handler`
              found  `&&users::Handler`
   = note: but, the lifetime must be valid for the static lifetime...
note: ...so that the type `[closure@src/main.rs:24:36: 24:69 self:&&users::Handler]` will meet its required lifetime bounds
  --> src/main.rs:24:18
   |
24 |                 .get("/users/:id", |request| self.load_user(request))
   |                  ^^^
```

Given this limitation, I don't understand how I can do anything useful with this crate. Am I missing something obvious? I would really appreciate the help.

**Comment by seanpianka** [comment]
Thanks for the solution, @rousan. Would there be any solutions besides making the state global? I have several different modules that will be creating sub-routers and added under scopes within a main router, and sharing this global static state across several modules seems like it will get messy quickly.

**Comment by rousan** [comment]
let me clarify, are you using a single state across all the modules or allocating a state for each module?

**Case-1:** You could use `lazy_static` crate to declare a central state in one place globally and use it across modules.
Define the state:
```rust
lazy_static! {
   pub static ref STATE: State= State::new();
}
```

**Case-2:** if you're trying to manage state per module, then just declare them as `private` global variable as follow:

```rust
lazy_static! {
   static ref DB_CLIENT: database::Client = database::Client::new();
}
```

**Comment by seanpianka** [comment]
My state is initialized at runtime, so it cannot be global static state. From my original example, The database client will be initialized with arguments from the command-line, which means your example with `database::Client::new()` does not match my use-case.

**Comment by seanpianka** [comment]
I'm looking for built-in support from routerify to accomplish something like the following to support accessing run-time initialized state from the handlers.

```rust
async fn main() {
    let addr = SocketAddr::from(([0,0,0,0], 3000));
    
    // My global state
    let state = Arc::new(RequestState { foo: 321 });
    
    // How can I make some sanity this onion1, onion2, onion3, onion4? Is it possible?
    let make_svc = make_service_fn(|_| {
        let onion1 = Arc::clone(&state);
        async move {
            Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
                let onion2 = Arc::clone(&onion1);
                async move {
                    handle_request(&*onion2, req).await
                }
            }))
        }
    });
    let server = Server::bind(&addr).serve(make_svc);
    if let Err(e) = server.await {
        eprintln!("server error: {}", e);
    }
}
```
Source: https://users.rust-lang.org/t/hyper-passing-a-state-reference-to-async-request-handler/36295/3

**Comment by seanpianka** [comment]
See this example from a different router which makes application state available in the handlers: https://github.com/kardeiz/reset-router/blob/master/reset-router/src/lib.rs#L232

**Comment by rousan** [comment]
Good point, I will add this support very soon.

**Comment by seanpianka** [comment]
Thank you! If at all possible, I'd like to help since this feature is blocking my ability to use this crate.

**Comment by rousan** [comment]
Will make a new release with the support today

**Comment by seanpianka** [comment]
Thanks again for the quick support on this! I can test it today

**Comment by rousan** [comment]
I just made a released `v1.1.3` by adding the above feature.

Doc: [Data and State Sharing](https://docs.rs/routerify/1.1.3/routerify/index.html#data-and-state-sharing).
